### PR TITLE
tests: mempool: Enhance tests to improve code coverage

### PR DESCRIPTION
**Pull request by ajaykish**
Validate allocation and free memory from system heap memory pool

Signed-off-by: Ajay Kishore <ajay.kishore@intel.com>

### CODE REVIEW

File: tests/kernel/mem_pool/sys_mem_pool/src/main.c [hunk 1]
```diff
@@ -19,10 +20,35 @@
...
+/**
+ * @see k_malloc(), k_free()
+ */
+#ifndef CONFIG_USERSPACE
+static void thread_handler(void *p1, void *p2, void *p3)
+{
+	char *ptr;
+
+	k_thread_system_pool_assign(k_current_get());
+	ptr = k_malloc(128);
```
Comment (andrewboie) [inline tests/kernel/mem_pool/sys_mem_pool/src/main.c]: k_malloc and k_thread_system_pool_assign have nothing to do with each other, what exactly are we testing here?

k_thread_system_pool_assign assigns a memory pool for z_thread_malloc, not k_malloc

also why is this test have a ifndef CONFIG_USERSPACE around it??

Comment (ajaykish) [reply tests/kernel/mem_pool/sys_mem_pool/src/main.c]: In this test case we are trying to validate k_thread_system_pool_assign() api.
We are assigning system heap resource to the current thread. And later trying to allocate and free memory from the system heap memory. 

As you suggested I have removed k_malloc and used z_thread_malloc() instead.

I was facing some shippable issue with qemu_x86, to resolve that issue I added "ifndef CONFIG_USERSPACE". I removed "ifndef CONFIG_USERSPACE" and not getting any shippable issue now.

File: include/kernel.h [hunk 2]
```diff
@@ -4191,6 +4191,17 @@ extern void k_free(void *ptr);
...
+ * @brief Allocate memory from system heap.
+ *
+ * Such memory must be released using k_free().
+ *
+ * @param size Amount of memory requested (in bytes).
+ *
+ * @return Address of the allocated memory if successful; otherwise NULL.
+ * @req K-HEAP-001
+ */
+extern void *z_thread_malloc(size_t size);
```
Comment (andrewboie) [inline include/kernel.h]: Why are you doing this????

This is an internal kernel API and it is already declared in kernel/include/kernel_internal.h!

File: tests/kernel/mem_pool/sys_mem_pool/src/main.c [hunk 3]
```diff
@@ -19,10 +20,29 @@
...
+K_SEM_DEFINE(end_sema, 0, 1);
+
+/**
+ * @see k_malloc(), k_free()
+ */
+static void thread_handler(void *p1, void *p2, void *p3)
+{
+	void *ptr;
+
+	k_thread_system_pool_assign(k_current_get());
```
Comment (andrewboie) [inline tests/kernel/mem_pool/sys_mem_pool/src/main.c]: This test case is to validate the sys_mem_pool APIs as defined in include/misc/mempool.h

Why are we shoving a test for k_thread_system_pool_assign() into this test case? It does not belong here.

**Review by tyrantlizard77 — CHANGES_REQUESTED** [review]

File: include/kernel.h (hunk 2 above)
Comment (ajaykish) [reply include/kernel.h]: @andrewboie Sorry, I did not check declaration in this file.
Included <kernel_internal.h>, in the test case to resolve implicit declaration of function error.

File: tests/kernel/mem_pool/sys_mem_pool/src/main.c (hunk 3 above)
Comment (ajaykish) [reply tests/kernel/mem_pool/sys_mem_pool/src/main.c]: @andrewboie Moved this test to tests/kernel/mem_pool/mem_pool_api/ folder.

File: tests/kernel/mem_pool/mem_pool_api/src/main.c [hunk 4]
```diff
@@ -19,6 +19,7 @@ extern void test_mpool_alloc_free_isr(void);
 extern void test_mpool_kdefine_extern(void);
 extern void test_mpool_alloc_size(void);
 extern void test_mpool_alloc_timeout(void);
+extern void test_sys_mem_pool_assign(void);
```
Comment (nashif) [inline tests/kernel/mem_pool/mem_pool_api/src/main.c]: change naming, this is to be confused with sys_mem_pool which is in another test.

**Review by nashif — CHANGES_REQUESTED** [review]

File: tests/kernel/mem_pool/mem_pool_api/src/main.c (hunk 4 above)
Comment (ajaykish) [reply tests/kernel/mem_pool/mem_pool_api/src/main.c]: @nashif Changed the test name to test_sys_heap_mem_pool_assign.

**Review by nashif — APPROVED** [review]

**Pull request merged**